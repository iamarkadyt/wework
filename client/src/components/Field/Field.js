import React from 'react'
import DropdownList from 'react-widgets/lib/DropdownList'
import DateTimePicker from 'react-widgets/lib/DateTimePicker'
import Multiselect from 'react-widgets/lib/Multiselect'
import Moment from 'moment'
import momentLocalizer from 'react-widgets-moment'
import './Field.scss'
import { node, string, func, object, bool, number, arrayOf, oneOfType, instanceOf } from 'prop-types'

Moment.locale('en')
momentLocalizer()

const field = ({
    // different fields require different props
    type,
    name, value, onChange, // all except submit and button
    placeholder, error, // all except checkbox, submit and button
    label, // checkbox, submit, button
    list, // list, multiselect
    rows, // textarea
    disabled, // date
    onClick, style, // button, linkButton, textarea
    inline, // container
    containerStyle, // container
    children
}) => {
    let field = null
    const commonProps = { name, id: name, value, onChange }

    switch (type) {
        case "textarea":
            field = <textarea
                {...commonProps}
                placeholder={placeholder || 'Say something...'}
                className="rw-widget-container rw-widget-input"
                style={{
                    padding: '.5rem',
                    fontSize: '1rem',
                    width: '100%',
                    resize: 'none',
                    ...style
                }}
                rows={rows} />
            break
        case "date":
            field = <DateTimePicker
                {...commonProps}
                placeholder={placeholder || 'Pick a date...'}
                time={false}
                format="MMM DD, YYYY"
                editFormat="M/DD/YYYY"
                disabled={disabled} />
            break
        case "list":
            field = <DropdownList
                {...commonProps}
                placeholder={placeholder || "Choose one..."}
                data={list} />
            break
        case "multiselect":
            field = <Multiselect
                {...commonProps}
                placeholder={placeholder || "Enter a few items..."}
                data={list} />
            break
        case "checkbox":
            field = <label htmlFor={name}>
                <input type="checkbox" {...commonProps} />
                &nbsp;&nbsp;{label}
            </label>
            break
        case "submit":
            field = <button
                type="submit"
                className="button-action">
                {label || children}
            </button>
            break
        case "button":
            field = <button
                className="button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        case "linkButton":
            field = <button
                className="link-button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        default:
            field = <input
                {...commonProps}
                placeholder={placeholder}
                className="rw-widget-container rw-widget-input"
                type={type} />
    }

    return (
        <div style={containerStyle}
            className={['Field-container', inline ? 'inline' : ''].join(' ')}>
            {label
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <label htmlFor={name}>{label}</label>}
            {field}
            {error
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <span className="error">{error}</span>}
        </div>
    )
}

field.propTypes = {
  type: string,
  name: string,
  value: oneOfType([string, arrayOf(string), bool, instanceOf(Date)]),
  onChange: func,
  placeholder: string,
  error: string,
  label: string,
  list: arrayOf(string),
  rows: oneOfType([string, number]),
  disabled: bool,
  onClick: func,
  style: object,
  inline: bool,
  containerStyle: object,
  children: oneOfType([arrayOf(node), node])
}

export default field
