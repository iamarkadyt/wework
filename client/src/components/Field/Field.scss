@import 'variables.scss';
.Field-container {
    &.inline {
        display: inline;
    }
    width: 100%;
    margin: 1rem 0 0 0;
    input[type=checkbox] {
        transform: scale(1.5);
    }
    .button,
    button[type=submit] {
        min-width: 6rem;
        border-radius: .2rem;
        font-size: 1.1rem;
        padding: .2rem 1rem;
        border-color: $color-button-border;
        background: $color-button-backround;
        color: $color-button-font;
        &:active {
            border-color: $color-button-active-border;
            background: $color-button-active-background;
            color: $color-button-active-font;
        }
        &:hover:not(:active) {
            border-color: $color-button-hover-border;
            background: $color-button-hover-backround;
            color: $color-button-hover-font;
        }
    }
    button[type=submit] {
        float: right;
    }
    .button-action {
        background-color: darken(cornflowerblue, 10%);
        border-color: darken(cornflowerblue, 10%);
        background: linear-gradient(to bottom, darken(cornflowerblue, 10%), darken(cornflowerblue, 20%));
        color: white;
        &:active {
            background: linear-gradient(to bottom, darken(cornflowerblue, 25%), darken(cornflowerblue, 15%));
        }
        &:hover:not(:active) {
            background: linear-gradient(to bottom, darken(cornflowerblue, 8%), darken(cornflowerblue, 18%));
        }
    }
    .link-button {
        background: none;
        border: none;
        font-weight: bold;
        font-size: 1.1rem;
        color: $color-font-secondary;
        &:hover {
            color: $color-font-main;
            cursor: pointer;
        }
    }
    input[type=text],
    input[type=password] {
        font-size: 1rem;
        padding: .6rem;
        width: 100%;
    }
    label {
        display: inline-block;
        padding: .2rem;
    }
    textarea {
        background-color: $color-field-background;
        border-color: $color-field-border;
        color: $color-field-font;
    }
    .error {
        display: inline-block;
        color: $color-font-error;
        padding: .2rem;
    }
}