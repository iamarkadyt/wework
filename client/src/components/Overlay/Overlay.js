import React from 'react'
import './Overlay.scss'

class Overlay extends React.Component {
  componentWillMount() {
    document.body.style = "overflow: hidden"
  }

  componentWillUnmount() {
    document.body.style = null
  }

  handleSubmit(e) {
    const { 
      onBackdropClick 
    } = this.props

    if (e.target.id === 'Overlay-container-backdrop') {
      onBackdropClick()
    }
  }

  render() {
    const { 
      children, 
      centered
    } = this.props

    return (
      <div id="Overlay-container-backdrop" className="Overlay-container" onClick={this.handleSubmit}>
        <div className={["Overlay-content", centered ? "Overlay-content-centered" : null].join(' ')}>
          {children}
        </div>
      </div>
    )
  }
}

export default Overlay
