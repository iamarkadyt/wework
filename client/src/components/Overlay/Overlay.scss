@import 'variables.scss';
.Overlay-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    min-height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: calc(#{$navigation_height} + 1rem) 0 2rem 0;
    .content {
        z-index: 1;
        background: white;
        padding: 0.3rem;
        border-radius: .3rem;
        box-shadow: $box_shadow_main;
    }
    .backdrop {
        position: fixed;
        top: 0;
        left: 0;
        width: 100vw;
        height: 100vh;
        background-color: rgba(0, 0, 0, 0.5);
    }
}