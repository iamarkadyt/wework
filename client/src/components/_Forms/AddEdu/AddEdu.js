import React from 'react'
import { withFormik, Form, Field } from 'formik'
import './AddEdu.css'

const addEdu = () => {
    return <div className="AddEdu-container">
        <Form>
            <label htmlFor="school">School:</label>
            <Field type="text" name="school" id="school" placeholder="UCLA" />

            <label htmlFor="degree">Degree:</label>
            <Field type="text" name="degree" id="degree" placeholder="Bachelor of Science" />

            <label htmlFor="fos">Field of Study:</label>
            <Field type="text" name="fos" id="fos" placeholder="Bioeconomics" />

            <label htmlFor="from">From:</label>
            <Field type="text" name="from" id="from" />

            <label htmlFor="to">To:</label>
            <Field type="text" name="to" id="to" />

            <label>
                <Field type="checkbox"
                    name="current"
                    id="current" /> Current?
            </label>

            <label htmlFor="description">Description:</label>
            <textarea name="description" id="description" rows="5" />

            <button type="submit">Add</button>
        </Form>
    </div>
}

export default withFormik({
    handleSubmit() {

    }
})(addEdu)