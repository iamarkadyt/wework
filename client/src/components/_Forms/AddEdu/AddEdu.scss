@mixin flexList {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;    
}

.AddEdu-container {
    width: 100%;
    height: 100vh;
    @include flexList();
    form {
        @include flexList();
    }
}