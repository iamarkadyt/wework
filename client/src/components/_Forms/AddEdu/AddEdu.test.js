import React from 'react'
import { _UnconnectedAddEdu as AddEdu, mapStateToProps } from './AddEdu'
import Field from '../../Field/Field'
import { shallow } from 'enzyme'
import Overlay from '../../Overlay/Overlay'
import cloneDeep from 'lodash.clonedeep'

describe('AddEdu', () => {
  let props
  let mountedAddEdu

  const getMockProps = () => {
    const actions = {
      addEducation: jest.fn()
    }

    const history = {
      goBack: jest.fn()
    }

    return cloneDeep({
      ...actions,
      history,
      errors: {}
    })
  }

  const comp = () => {
    if (!mountedAddEdu) {
      mountedAddEdu = shallow(<AddEdu {...props} />)
    }
    return mountedAddEdu
  }

  beforeEach(() => {
    props = getMockProps()
    mountedAddEdu = undefined
  })

  describe('props passing', () => {
    describe('Overlay', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'onBackdropClick',
          'children'
        ]
        expect(Object.keys(comp().find(Overlay).props())).toStrictEqual(expectedKeys)
      })
      
      it('receives correct [onBackdropClick] prop', () => {
        expect(comp().find(Overlay).prop('onBackdropClick')).toEqual(comp().instance().handleDismiss)
      })
    })

    describe('form', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'className',
          'onSubmit',
          'children'
        ]
        expect(Object.keys(comp().find('form').props())).toStrictEqual(expectedKeys)
      })
      
      it('receives correct [onSubmit] prop', () => {
        expect(comp().find('form').prop('onSubmit')).toEqual(expect.any(Function))
      })
    })

    describe('school Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
          'placeholder'
        ]
        expect(Object.keys(comp().find(Field).find('[name="school"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="school"]').prop('type')).toBe('text')
      })

      it('receives state.school into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ school: state })
        expect(comp().find(Field).find('[name="school"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.school is not defined', () => {
        comp().setState({ school: undefined })
        expect(comp().find(Field).find('[name="school"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="school"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="school"]').prop('label')).toBe('School:')
      })

      it('receives props.errors.school into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.school = msg
        expect(comp().find(Field).find('[name="school"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.school into [error] prop if former is defined', () => {
        delete props.errors.school
        expect(comp().find(Field).find('[name="school"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [placeholder] prop', () => {
        expect(comp().find(Field).find('[name="school"]').prop('placeholder')).toBe('UCLA')
      })
    })

    describe('fieldOfStudy Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
          'placeholder'
        ]
        expect(Object.keys(comp().find(Field).find('[name="fieldOfStudy"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('type')).toBe('text')
      })

      it('receives state.fieldOfStudy into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ fieldOfStudy: state })
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.fieldOfStudy is not defined', () => {
        comp().setState({ fieldOfStudy: undefined })
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('label')).toBe('Field of Study:')
      })

      it('receives props.errors.fieldOfStudy into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.fieldOfStudy = msg
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.fieldOfStudy into [error] prop if former is defined', () => {
        delete props.errors.fieldOfStudy
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [placeholder] prop', () => {
        expect(comp().find(Field).find('[name="fieldOfStudy"]').prop('placeholder')).toBe('Bioeconomics')
      })
    })

    describe('degree Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'list',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="degree"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="degree"]').prop('type')).toBe('list')
      })

      it('receives state.degree into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ degree: state })
        expect(comp().find(Field).find('[name="degree"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.degree is not defined', () => {
        comp().setState({ degree: undefined })
        expect(comp().find(Field).find('[name="degree"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="degree"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="degree"]').prop('label')).toBe('Degree:')
      })

      it('receives correct [list] prop', () => {
        expect(comp().find(Field).find('[name="degree"]').prop('list')).toEqual(expect.any(Array))
      })

      it('receives props.errors.degree into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.degree = msg
        expect(comp().find(Field).find('[name="degree"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.degree into [error] prop if former is defined', () => {
        delete props.errors.degree
        expect(comp().find(Field).find('[name="degree"]').prop('error')).not.toBeDefined()
      })
    })

    describe('[from] Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="from"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="from"]').prop('type')).toBe('date')
      })

      it('receives state.from into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ from: state })
        expect(comp().find(Field).find('[name="from"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.from is not defined', () => {
        comp().setState({ from: undefined })
        expect(comp().find(Field).find('[name="from"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="from"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="from"]').prop('label')).toBe('From:')
      })

      it('receives props.errors.from into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.from = msg
        expect(comp().find(Field).find('[name="from"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.from into [error] prop if former is defined', () => {
        delete props.errors.from
        expect(comp().find(Field).find('[name="from"]').prop('error')).not.toBeDefined()
      })
    })

    describe('[to] Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'disabled',
          'label',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="to"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="to"]').prop('type')).toBe('date')
      })

      it('receives state.to into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ to: state })
        expect(comp().find(Field).find('[name="to"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.to is not defined', () => {
        comp().setState({ to: undefined })
        expect(comp().find(Field).find('[name="to"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="to"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives state.current into [disabled] prop if former is defined', () => {
        const state = false
        comp().setState({ current: state })
        expect(comp().find(Field).find('[name="to"]').prop('disabled')).toBe(state)
      })

      it('receives undefined into [disabled] prop if state.current is not defined', () => {
        comp().setState({ current: undefined })
        expect(comp().find(Field).find('[name="to"]').prop('disabled')).not.toBeDefined()
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="to"]').prop('label')).toBe('To:')
      })

      it('receives props.errors.to into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.to = msg
        expect(comp().find(Field).find('[name="to"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.to into [error] prop if former is defined', () => {
        delete props.errors.to
        expect(comp().find(Field).find('[name="to"]').prop('error')).not.toBeDefined()
      })
    })

    describe('[current] Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
        ]
        expect(Object.keys(comp().find(Field).find('[name="current"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="current"]').prop('type')).toBe('checkbox')
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="current"]').prop('label')).toBe('Current?')
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="current"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives state.current into [value] prop if former is defined', () => {
        const state = true
        comp().setState({ current: state })
        expect(comp().find(Field).find('[name="current"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.current is not defined', () => {
        comp().setState({ current: undefined })
        expect(comp().find(Field).find('[name="current"]').prop('value')).not.toBeDefined()
      })
    })

    describe('description Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'rows',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="description"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('type')).toBe('textarea')
      })

      it('receives state.description into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ description: state })
        expect(comp().find(Field).find('[name="description"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.description is not defined', () => {
        comp().setState({ description: undefined })
        expect(comp().find(Field).find('[name="description"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('label')).toBe('Description')
      })

      it('receives props.errors.description into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.description = msg
        expect(comp().find(Field).find('[name="description"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.description into [error] prop if former is defined', () => {
        delete props.errors.description
        expect(comp().find(Field).find('[name="description"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [rows] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('rows')).toBe('2')
      })
    })

    describe('submit button', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'label',
        ]
        expect(Object.keys(comp().find(Field).find('[name="submit"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="submit"]').prop('type')).toBe('submit')
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="submit"]').prop('label')).toBe('Add')
      })
    })

    describe('cancel button', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'label',
          'onClick'
        ]
        expect(Object.keys(comp().find(Field).find('[name="cancel"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="cancel"]').prop('type')).toBe('button')
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="cancel"]').prop('label')).toBe('Cancel')
      })

      it('receives correct [onClick] prop', () => {
        expect(comp().find(Field).find('[name="cancel"]').prop('onClick')).toEqual(expect.any(Function))
      })
    })
  })

  describe('functions', () => {
    describe('handleDismiss()', () => {
      it('calls history.goBack() when invoked', () => {
        comp().instance().handleDismiss()
        expect(props.history.goBack).toHaveBeenCalled()
      })
    })
  })

  describe('interaction', () => {
    describe('form submission', () => {
      let mockedEvent

      beforeEach(() => {
        mockedEvent = { preventDefault: jest.fn() }
      })

      it('invokes preventDefault()', () => {
        comp().find('form').simulate('submit', mockedEvent)
        expect(mockedEvent.preventDefault).toHaveBeenCalled()
      })

      it('invokes addExperience()', () => {
        comp().find('form').simulate('submit', mockedEvent)
        expect(props.addEducation).toHaveBeenCalled()
      })

      it('invokes addExperience() with correct args', () => {
        comp().find('form').simulate('submit', mockedEvent)
        expect(props.addEducation).toHaveBeenCalledWith(comp().state(), comp().instance().handleDismiss)
      })
    })

    describe('Fields', () => {
      let mockedEvent
      let value = 'some value'

      beforeEach(() => {
        mockedEvent = { 
          target: { value }, 
          preventDefault: jest.fn()
        }
      })
      
      describe('school', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="school"]').simulate('change', mockedEvent)
          expect(comp().state('school')).toBe(value)
        })
      })

      describe('fieldOfStudy', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="fieldOfStudy"]').simulate('change', mockedEvent)
          expect(comp().state('fieldOfStudy')).toBe(value)
        })
      })

      describe('degree', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="degree"]').simulate('change', value)
          expect(comp().state('degree')).toBe(value)
        })
      })

      describe('from', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="from"]').simulate('change', value)
          expect(comp().state('from')).toBe(value)
        })
      })

      describe('to', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="to"]').simulate('change', value)
          expect(comp().state('to')).toBe(value)
        })
      })

      describe('current', () => {
        it('[onChange] function works as expected', () => {
          const value = false
          mockedEvent.target.checked = value
          comp().find(Field).find('[name="current"]').simulate('change', mockedEvent)
          expect(comp().state('current')).toBe(value)
        })
      })

      describe('description', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="description"]').simulate('change', mockedEvent)
          expect(comp().state('description')).toBe(value)
        })
      })

      describe('cancel', () => {
        describe('[onClick]', () => {
          it('invokes preventDefault()', () => {
            comp().find(Field).find('[name="cancel"]').simulate('click', mockedEvent)
            expect(mockedEvent.preventDefault).toHaveBeenCalled()
          })

          it('invokes handleDismiss()', () => {
            const mockFn = jest.fn()
            comp().instance().handleDismiss = mockFn
            comp().find(Field).find('[name="cancel"]').simulate('click', mockedEvent)
            expect(mockFn).toHaveBeenCalled()
          })
        })
      })
    })
  })

  describe('mapStateToProps', () => {
    it('returns expected object', () => {
      const state = {
        err: { 
          formErrors: {
            email: 'Email is wrong' 
          }
        }
      }

      const expectedObject = {
        errors: state.err.formErrors
      }

      expect(mapStateToProps(state)).toEqual(expectedObject)
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})
