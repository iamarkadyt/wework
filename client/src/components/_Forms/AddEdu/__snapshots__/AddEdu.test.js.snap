// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEdu matches snapshot 1`] = `
<Overlay
  onBackdropClick={[Function]}
>
  <form
    className="AddEdu-container"
    onSubmit={[Function]}
  >
    <h1>
      Add Education
    </h1>
    <field
      label="School:"
      name="school"
      onChange={[Function]}
      placeholder="UCLA"
      type="text"
    />
    <field
      label="Field of Study:"
      name="fieldOfStudy"
      onChange={[Function]}
      placeholder="Bioeconomics"
      type="text"
    />
    <field
      label="Degree:"
      list={
        Array [
          "Professional Certificate",
          "Undergraduate Degrees",
          "Transfer Degree",
          "Associate Degree",
          "Bachelor Degree",
          "Graduate Degrees",
          "Master Degree",
          "Doctoral Degree",
          "Professional Degree",
          "Specialist Degree",
        ]
      }
      name="degree"
      onChange={[Function]}
      type="list"
    />
    <field
      label="From:"
      name="from"
      onChange={[Function]}
      type="date"
    />
    <field
      label="To:"
      name="to"
      onChange={[Function]}
      type="date"
    />
    <field
      label="Current?"
      name="current"
      onChange={[Function]}
      type="checkbox"
    />
    <field
      label="Description"
      name="description"
      onChange={[Function]}
      rows="2"
      type="textarea"
    />
    <field
      label="Add"
      name="submit"
      type="submit"
    />
    <field
      label="Cancel"
      name="cancel"
      onClick={[Function]}
      type="button"
    />
  </form>
</Overlay>
`;
